### PR TITLE
Integ tests

### PR DESCRIPTION
**Pull request by Vahila**
Integrated MATLAB with Jenkins Test cases

### CODE REVIEW

**Comment by Vahila** [comment]
The failed test case is related to Matrix build. The test case checks that in Matrix build if the specified MATLAB Version is not available, then the build should fail. But during the build, if the specified Version is not available, the test cases are run on the available version and the build is Successful. Hence the test case is failing.

@SangaviJayagopi @nbhoski @abhargav-mw Please review the changes

**Comment by SangaviJayagopi** [comment]
@nbhoski @abhargav-mw Along with the review for these changes, would also like to know your comments on merging these changes into the master regarding maintaining these tests with the IntegTests branch or master branch.

I prefer keeping them in the Integ tests branch and whenever you create a QE branch for verification, we could pull those changes to the Integ tests branch, run the tests with those changes and approve changes on the QE branch based on build in the Integ tests branch.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -3,17 +3,7 @@
...
-    linux:
-      imageName: 'ubuntu-16.04'
-    mac:
-      imageName: 'macOS-10.15'
-    windows:
-      imageName: 'vs2017-win2016'
-
-pool:
-  vmImage: $(imageName)
+pool: Default
```
Comment (SangaviJayagopi) [inline azure-pipelines.yml:6]: Think about matrix build using self-hosted agents. Also, I guess the name of the pool is 'Default' and a comment about what OS agent is used would be more informative

File: src/integ-test/java/com/mathworks/ci/RunMATLABCommandInteg.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    @After
+    public void testTearDown() {
+        this.project = null;
+        this.scriptBuilder = null;
+    }
+
+    private String getMatlabroot() throws URISyntaxException {
+        String ML_version = TestData.getPropValues("matlab.version");
```
Comment (SangaviJayagopi) [inline src/integ-test/java/com/mathworks/ci/RunMATLABCommandInteg.java:57]: Is the version of MATLAB specified here  also taken as input for MATLAB windows/linux root path?

File: src/integ-test/java/com/mathworks/ci/RunMATLABTestsInteg.java [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+    private RunMatlabTestsBuilder testBuilder;
+
+
+    @Rule
+    public JenkinsRule jenkins = new JenkinsRule();
+
+    @Before
+    public void testSetup() throws IOException {
+
+        this.project = jenkins.createFreeStyleProject();
```
Comment (SangaviJayagopi) [inline src/integ-test/java/com/mathworks/ci/RunMATLABTestsInteg.java:47]: The test cases here are based on the FreeStyleProject configuration. Could start thinking about how to add test cases for a Pipeline project too.

File: src/integ-test/java/com/mathworks/ci/RunMATLABTestsInteg.java [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        jenkins.assertBuildStatus(Result.FAILURE, build2);
+    }
+    
+    /*
+     * Test to verify if Matrix build passes (mock MATLAB).
+     */
+    @Test
+    public void verifyMatrixBuildPasses() throws Exception {
+        MatrixProject matrixProject = jenkins.createProject(MatrixProject.class);
+        Axis axes = new Axis("VERSION", "R2020a", "R2020a");
```
Comment (SangaviJayagopi) [inline src/integ-test/java/com/mathworks/ci/RunMATLABTestsInteg.java:246]: Are the two versions mentioned here supposed to be same?

**Review by SangaviJayagopi — COMMENTED** [review]
In general I would like to comment about:

1. Other test cases to be added in case of a Pipeline project or additional integration tests. 
2. Scheduling builds in azure for a daily/weekly build.

File: azure-pipelines.yml (hunk 1 above)
Comment (Vahila) [reply azure-pipelines.yml:6]: A Windows Self hosted machine is used for the build. 
I will add the comments in the file.

File: src/integ-test/java/com/mathworks/ci/RunMATLABTestsInteg.java (hunk 4 above)
Comment (Vahila) [reply src/integ-test/java/com/mathworks/ci/RunMATLABTestsInteg.java:246]: No, the versions need not be same but they have to be installed on the machine. I will correct the test case with different versions.

File: azure-pipelines.yml (hunk 1 above)
Comment (nbhoski) [reply azure-pipelines.yml:6]: I think we need to discuss on whether we could use these self-hosted agents and if we don then definitely we need to run it on several OS as we do currently. changing this and merging it to master will affect all our current dev process and checks that it does on different os.  I suggest reserving this change to be merged in the master branch

**Review by nbhoski — CHANGES_REQUESTED** [review]
I see integ test failed during verify 

--- maven-failsafe-plugin:2.19.1:verify (integration-tests) @ matlab ---
2020-10-12T10:30:51.2480140Z [INFO] ------------------------------------------------------------------------
2020-10-12T10:30:51.2480659Z [INFO] BUILD FAILURE
2020-10-12T10:30:51.2482176Z [INFO] ------------------------------------------------------------------------
2020-10-12T10:30:51.2496791Z [INFO] Total time:  11:57 min
2020-10-12T10:30:51.2498316Z [INFO] Finished at: 2020-10-12T16:00:51+05:30
2020-10-12T10:30:51.2500409Z [INFO] ------------------------------------------------------------------------
2020-10-12T10:30:51.2511140Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:2.19.1:verify (integration-tests) on project matlab: There are test failures.

File: README.md [hunk 5]
```diff
@@ -15,6 +15,7 @@ To learn how to use the plugin in testing workflows, see [Examples](/examples/).
 ## Contact Us
 If you have any questions or suggestions, please contact MathWorks&reg; at [continuous-integration@mathworks.com](mailto:continuous-integration@mathworks.com).
```
Comment (nbhoski) [inline README.md:17]: Why is this file changed ?

**Review by nbhoski — CHANGES_REQUESTED** [review]
I see The branch is created and merge request in Master. As per the process i would request to create a resh PR which request merge into dev_main and also the integ ranch is created out of dev_main

**Comment by Vahila** [comment]
Thank you for reviewing @SangaviJayagopi @nbhoski  @abhargav-mw. I will close this PR and create new one with updated test cases and edits.
